docs/replication.md

```diff
@@ -0,0 +1,145 @@
+
+In both case, we need to support
+
+- `PUT    :both/_local/:revdocid` to store the current sequence number.
+
+# Stack Sync API exploration
+
+
+## Easy part: 1 db/doctype on stack AND remote, no binaries
+We *just* need to implement the routes described above by proxying to the underlying couchdb database.
+
+```javascript
+var db = new PouchDB("my-local-contacts")
+db.replicate.from('https://bob.cozycloud.cc/data/contacts')
+db.replicate.to('https://bob.cozycloud.cc/data/contacts')
+```
+
+To suport this we need to :
+- Proxy `_changes` route with since, limit, feed=normal. Refuse all filter parameters. [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/changes.html)
+- Add support of `open_revs` query parameter to `GET /data/:doctype/:docid` [(Doc) ](http://docs.couchdb.org/en/2.0.0/api/document/common.html?highlight=open_revs#get--db-docid)
+- Proxy the `_revs_diff` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/misc.html#db-revs-diff) and `_bulk_docs` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/bulk-api.html) routes
+- Have `_ensure_full_commit` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/compact.html#db-ensure-full-commit) returns 201
+
+This will cover documents part of the Devices use case.
+
+With a way to list doctype, this could also cover Backup and Moving.
+
+
+## Continuous replication
+
+It is impossible to implement it by simply proxying to couchdb (see unlimited inactive users). But neither the Device nor the Backup option need open continuous replication.
+
+The only use case for which we might want it is sharing. Sharing has the advantage of being (at first) stack to stack. So we might imagine having the source of event "ping" the remote through a separate route.
+
+**Conclusion** : no continuous replication except for a few special case (collaborative edit)
+
+
+## Realtime
+
+One of the cool feature of cozy apps was how changes are sent to the client in realtime through websocket. However this have a cost : every cozy and every apps open in a browser, even while not used keep one socket open on the server.
+
+Several technology can be used for this : Websocket, SSE or COMET like long-polling. Goroutines makes all solution similar performances wise. COMET is a hack, websocket seems more popular and tested (x/net/websocket vs html5-sse-example). SSE has some limitations (can't set header...)
+
+Depending on benchmarking, we can do some optimization on the feed :
+- close feeds when the user is not on screen
+- multiplex different applications' feed, so each open cozy will only use one socket to the server. And then use a [SharedWorker](https://developer.mozilla.org/en/docs/Web/API/SharedWorker) or postMessage to pass the events to each application javascript context (either iframe or browser tabs).
```

Comment:
The apps won't be on the same domains and we will make them not embeddable in iframes for security reasons. I don't think this let us some options to mutualize connections between apps. But we can see that later, it's only optimizations.